File: probs/two-pointer/11.container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int l = 0;
        int r = height.length - 1;
        int answer = 0;
        while (l < r) {
            int smaller = height[l] > height[r] ? r: l;
            int area = height[smaller] * (r - l);

            if (area > answer) {
                answer = area;
            }
            if (height[l] > height[r]) {
                r--;
            } else {
                l++;
            }
        }

        return answer;
    }
}

File: top-interview-150/11.container-with-most-water.kt
class Solution {
    fun maxArea(height: IntArray): Int {
        var left = 0
        var right = height.size - 1
        var answer = 0

        while (left < right) {
            val area = minOf(height[left], height[right]) * (right - left)
            answer = maxOf(answer, area)

            // Core insight: move the pointer with smaller height
            // Why? The shorter bar limits the area. Moving the taller bar
            // will only decrease width without possibility of increasing area.
            // Moving the shorter bar gives us a chance to find a taller bar.
            if (height[left] < height[right]) {
                left++
            } else {
                right--
            }
        }

        return answer
    }
}

